File: templates/problems/03_sliding_window/05_min_window_substring/typescript/solution.ts
/**
 * Find the minimum window in s that contains all characters of t.
 * Time: O(n), Space: O(1) - fixed character set
 */
export function solveMinWindow(s: string, t: string): string {
  if (t.length === 0) {
    return '';
  }

  const tCount: number[] = new Array(128).fill(0);
  const windowCount: number[] = new Array(128).fill(0);
  for (let i = 0; i < t.length; i++) {
    tCount[t.charCodeAt(i)]++;
  }

  let need = 0;
  for (let i = 0; i < 128; i++) {
    if (tCount[i] > 0) need++;
  }

  let have = 0;
  let bestLen = s.length + 1;
  let bestStart = 0;
  let left = 0;

  for (let right = 0; right < s.length; right++) {
    const c = s.charCodeAt(right);
    windowCount[c]++;
    if (tCount[c] > 0 && windowCount[c] === tCount[c]) {
      have++;
    }

    while (have === need) {
      if (right - left + 1 < bestLen) {
        bestLen = right - left + 1;
        bestStart = left;
      }
      const lc = s.charCodeAt(left);
      windowCount[lc]--;
      if (tCount[lc] > 0 && windowCount[lc] < tCount[lc]) {
        have--;
      }
      left++;
    }
  }

  if (bestLen > s.length) {
    return '';
  }
  return s.substring(bestStart, bestStart + bestLen);
}

File: templates/problems/03_sliding_window/05_min_window_substring/python/solution.py
def solve_min_window(s: str, t: str) -> str:
    """Find the minimum window in s that contains all characters of t.

    Time: O(n), Space: O(1) - fixed character set
    """
    if len(t) == 0:
        return ""

    t_count = [0] * 128
    window_count = [0] * 128
    for ch in t:
        t_count[ord(ch)] += 1

    need = sum(1 for c in t_count if c > 0)

    have = 0
    best_len = len(s) + 1
    best_start = 0
    left = 0

    for right in range(len(s)):
        c = ord(s[right])
        window_count[c] += 1
        if t_count[c] > 0 and window_count[c] == t_count[c]:
            have += 1

        while have == need:
            if right - left + 1 < best_len:
                best_len = right - left + 1
                best_start = left
            lc = ord(s[left])
            window_count[lc] -= 1
            if t_count[lc] > 0 and window_count[lc] < t_count[lc]:
                have -= 1
            left += 1

    if best_len > len(s):
        return ""
    return s[best_start : best_start + best_len]
